src/dxvk/dxvk_openvr.cpp

```diff
@@ -76,31 +89,137 @@ namespace dxvk {
   void VrInstance::initDeviceExtensions(const DxvkInstance* instance) {
     std::lock_guard<std::mutex> lock(m_mutex);
 
-    if (!m_compositor || m_initializedDevExt)
+    if (m_no_vr || (!m_vr_key && !m_compositor) || m_initializedDevExt)
       return;
     
     for (uint32_t i = 0; instance->enumAdapters(i) != nullptr; i++) {
       m_devExtensions.push_back(this->queryDeviceExtensions(
-        instance->enumAdapters(i)->handle()));
+        instance->enumAdapters(i)));
     }
 
     m_initializedDevExt = true;
     this->shutdown();
   }
 
+  bool VrInstance::waitVrKeyReady() const {
+    DWORD type, value, wait_status, size;
+    LSTATUS status;
+    HANDLE event;
+
+    size = sizeof(value);
+    if ((status = RegQueryValueExA(m_vr_key, "state", NULL, &type, (BYTE *)&value, &size)))
+    {
+        Logger::err(str::format("OpenVR: could not query value, status ", status));
+        return false;
+    }
+    if (type != REG_DWORD)
+    {
+        Logger::err(str::format("OpenVR: unexpected value type ", type));
+        return false;
+    }
+
+    if (value)
+        return value == 1;
+
+    event = CreateEventA( NULL, FALSE, FALSE, NULL );
+    while (1)
+    {
+        if (RegNotifyChangeKeyValue(m_vr_key, FALSE, REG_NOTIFY_CHANGE_LAST_SET, event, TRUE))
+        {
+            Logger::err("Error registering registry change notification");
+            goto done;
+        }
+        size = sizeof(value);
+        if ((status = RegQueryValueExA(m_vr_key, "state", NULL, &type, (BYTE *)&value, &size)))
+        {
+            Logger::err(str::format("OpenVR: could not query value, status ", status));
+            goto done;
+        }
+        if (value)
+            break;
+        while ((wait_status = WaitForSingleObject(event, 1000)) == WAIT_TIMEOUT)
```

Comment:
Do we have any sort of guarantee that setting the registry key can never go wrong and that we'll never have stale entries in the registry even when something crashed during a previous launch? If not, we really shouldn't be creating an infinite loop here.

Reply:
> Why does the current code not work outside of some potential race conditions if the app initializes both at the same time? We shouldn't touch initialization unless `getGenericInterface` fails to return a compositor.

Yes, you are right, assuming DXVK gets the same higher level openvr_api.dll loader as the game is using (which stores the global vrclient's interface pointer during initialization and will succeed the interface query if the initialization has already been performed through VR_InitInternal[2] from this dll). That must be usually the case in practice, the other cases are more of theoretical concern and I probably shouldn't have mentioned this part.

The real issue which I currently saw affecting games is the extra initialization and teardown of vrclient done in the game process.

> 
> I'm not really a big fan of this solution either since it heavily depends on external factors working correctly, which also seems more than a little bit fragile.

Yes, doing the same VR client init / tear down outside of game requests is still not ideal (as vrclient links some things to steam game id also), but looks less fragile than doing the same in the game process.

> Do we have any sort of guarantee that setting the registry key can never go wrong and that we'll never have stale entries in the registry even when something crashed during a previous launch? If not, we really shouldn't be creating an infinite loop here.

The guarantee from any leftovers from the previous launch is using the volatile keys, they are not stored anywhere persistently and do not exist after Proton prefix shutdown, so upon new start the HKCU\\Software\\Wine\\VR key never exists until created by Proton's steam.exe. Unless someone sets it manually of course, but I believe we already have enough ways of breaking things manually like that anyway.

The more concerning part is probably the extension query process with VR init in initialize_vr_data() in steam.exe (in the attached patch). But I believe it is effectively equivalent to what DXVK currently does, so if something is going to hang there and leave status '0' the same seems to have equal chances to happen on the same route DXVK. If SteamVR is not running or libraries are not available we should be setting the ~0u status quickly, and it is rather hard to guess what can go wrong on this path unless some fundamental things (like setting registry key) are not working in Proton prefix.

I thought of adding some sort of timeout, but that timeout must be rather big. FWIW, with the way currently in DXVK the initialization takes ~2 sec in many cases when SteamVR is running with Steam's vrclient log complaining about timeout waiting for valid appkey after 2 sec. I am avoiding that now in steam.exe VR init, but not entirely sure such thing cannot happen under some different conditions. So my reasoning was that introducing the timeout will be only hiding some problems if they are going to happen. If you think it worth it though I don't mind adding such.